modules/sync/unique_queue.go

```diff
@@ -48,15 +79,8 @@ func (q *UniqueQueue) Exist(id interface{}) bool {
 // AddFunc adds new instance to the queue with a custom runnable function,
 // the queue is blocked until the function exits.
 func (q *UniqueQueue) AddFunc(id interface{}, fn func()) {
-	q.AddCtxFunc(context.Background(), id, fn)
-}
-
-// AddCtxFunc adds new instance to the queue with a custom runnable function,
-// the queue is blocked until the function exits. If the context is done before
-// the id is added to the queue it will not be added and false will be returned.
-func (q *UniqueQueue) AddCtxFunc(ctx context.Context, id interface{}, fn func()) bool {
 	if q.Exist(id) {
-		return true
+		return
 	}
```

Comment:
There's a potential race condition when two goroutines attempt to add functions with the same ID and neither exists before this line, but they keep going end are both ran below. I know we've tried to avoid duplicating Exist() because of the lock, but I think the check for existance should be done inside the same lock as the rest of the function.

Reply:
Damn. I didn't actually check that this code was syncing correctly.  I need to stop assuming that this has been done correctly.

Reply:
OK I've decided do a double check as the pool has a `RWMutex`.

Reply:
(This has made me realise this code needs to be entirely eaten by the queue stuff in the WIP)

Reply:
> 
> 
> OK I've decided do a double check as the pool has a `RWMutex`.

Mutexes are pretty fast devices. The chances are high that we'll have to enter the write lock anyway. so I'd skip the read lock. There's no harm in "write locking" without doing an actual write. 😁

Reply:
I mean: I'd skip the read lock. 😄